src/SyGuS.ml

```diff
@@ -111,28 +118,21 @@ let parse_sexps (sexps : Sexp.t list) : t =
                     in if List.mem !variables new_var ~equal:(fun x y -> String.equal (fst x) (fst y))
                        then raise (Parse_Exn ("Multiple declarations of variable " ^ (fst new_var)))
                        else variables := new_var :: !variables
-              | List [ (Atom "declare-fun") ; name ; args ; rtype ]
-                -> if args <> List [] then raise (Parse_Exn "Only nullary function (i.e. variable) declarations supported.") else
-                   let new_var = parse_variable_declaration (List [name ; rtype])
-                    in if List.mem !variables new_var ~equal:(fun x y -> String.equal (fst x) (fst y))
-                       then raise (Parse_Exn ("Multiple declarations of variable " ^ (fst new_var)))
-                       else variables := new_var :: !variables
-              | List ( (Atom "declare-primed-var") :: sexps )
-                -> let _var, _type = parse_variable_declaration (List sexps)
-                    in if List.mem !variables (_var, _type) ~equal:(fun x y -> String.equal (fst x) (fst y))
-                       then raise (Parse_Exn ("Multiple declarations of variable " ^ _var))
-                       else variables := (_var, _type) :: (_var ^ "!", _type) :: !variables
               | List ( (Atom "define-fun") :: func_sexps )
                 -> let (func, fconsts) = parse_define_fun func_sexps
                     in if List.mem !funcs func ~equal:(fun x y -> String.equal x.name y.name)
                        (* FIXME: SyGuS format allows overloaded functions with different signatures *)
                        then raise (Parse_Exn ("Multiple definitions of function " ^ func.name))
-                       else funcs := func :: !funcs ; consts := fconsts @ !consts
+                       else funcs := func :: !funcs ; consts := fconsts @ !consts ;
+                       (if String.equal func.name "trans_fun"
+                        then variables := func.args)
               | List [ (Atom "inv-constraint") ; (Atom _invf_name) ; (Atom _pref_name)
                                                ; (Atom _transf_name) ; (Atom _postf_name) ]
                 -> pref_name := _pref_name ; transf_name := _transf_name ; postf_name := _postf_name
-                 ; if not (String.equal !invf_name _invf_name)
-                   then raise (Parse_Exn ("Invariant function [" ^ _invf_name ^ "] not declared"))
+                 ; (if not (String.equal !invf_name _invf_name)
+                   then raise (Parse_Exn ("Invariant function [" ^ _invf_name ^ "] not declared")))
+                 ; if not (check_arg_order !variables !funcs)
+                   then raise (Parse_Exn ("Argument order not consistent for declared functions"))
```

Comment:
A SyGuS file can have other auxiliary functions besides our pre, trans, post, inv functions. We shouldn't be checking `check_arg_order` for all functions -- just those 4.

And If `check_arg_order` succeeds then you should update `variables`. To do that you would have to find the function corresponding to `_transf_name` within `funcs` and update variables with the function's `args` field.

Reply:
Could a variable ever be declared but not used in the trans function? In the SyGuS document it says that `inv-constraint` cannot accept global variables, so do we even need to support `declare-var` anymore?

Reply:
Quoting the specification document,
>  This [`declare-primed-var`] command did not provide any benefit to invariant synthesis problems, since constraints declared via inv-constraint do not accept global variables.

`inv-constraint` implicitly assumes that inv, pre, post and trans are defined on the same sequence of variables. But any of the inv, pre, post and trans functions themselves can depend on global variables.

But for now, could you remove the code that handles `declare-var` and raise an exception there that we do not currently support it? I will take a pass later and add some code to keep track of variables in various contexts.

Reply:
Ok, now I do all checks once `inv-constraint` is defined, getting the list of variables from the trans function, so `declare-var` is no longer supported.